### PR TITLE
Add `AsFd` implementations for stdio lock types on WASI.

### PR DESCRIPTION
**Pull request by sunfishcode**
This mirrors the implementations on Unix platforms, and also mirrors the existing `AsRawFd` impls.

This is similar to #100892, but is for the `*Lock` types.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
@bors r-

Are you sure this actually compiles on wasi? I don't see where StdinLock (or the other types) is imported in that module, maybe just missing something obvious.

**Comment by sunfishcode** [comment]
@Mark-Simulacrum Ah, you're right. That's now fixed, and as you mention, the compiler required stability attributes. I marked them "stable", using the io_safety feature, on the theory that implementing `AsFd` for these types is something that logically belongs to that feature, and was only omitted as an oversight.

**Comment by Mark-Simulacrum** [comment]
Hm, ok. If these are stable surface area then they need a libs-api FCP; r? @joshtriplett

**Comment by joshtriplett** [comment]
@rfcbot merge

File: library/std/src/sys/wasi/stdio.rs [hunk 1]
```diff
@@ -30,6 +30,14 @@ impl AsFd for Stdin {
     }
 }
 
+#[stable(feature = "io_safety", since = "1.63.0")]
+impl<'a> AsFd for io::StdinLock<'a> {
+    #[inline]
+    fn as_fd(&self) -> BorrowedFd<'_> {
+        unsafe { BorrowedFd::borrow_raw(0) }
```
Comment (hamza1311) [inline library/std/src/sys/wasi/stdio.rs:37]: Why exactly is this correct? A safety comment here would be great to have for someone stumbling upon this without having prior knowledge of the API and WASI spec

Comment (joshtriplett) [inline library/std/src/sys/wasi/stdio.rs:37]:
```suggestion
        // SAFETY: user code should not close stdin out from under the standard library
        unsafe { BorrowedFd::borrow_raw(0) }
```

File: library/std/src/sys/wasi/stdio.rs [hunk 2]
```diff
@@ -65,6 +73,14 @@ impl AsFd for Stdout {
     }
 }
 
+#[stable(feature = "io_safety", since = "1.63.0")]
+impl<'a> AsFd for io::StdoutLock<'a> {
+    #[inline]
+    fn as_fd(&self) -> BorrowedFd<'_> {
+        unsafe { BorrowedFd::borrow_raw(1) }
```
Comment (joshtriplett) [inline library/std/src/sys/wasi/stdio.rs:80]:
```suggestion
        // SAFETY: user code should not close stdout out from under the standard library
        unsafe { BorrowedFd::borrow_raw(1) }
```

File: library/std/src/sys/wasi/stdio.rs [hunk 3]
```diff
@@ -103,6 +119,14 @@ impl AsFd for Stderr {
     }
 }
 
+#[stable(feature = "io_safety", since = "1.63.0")]
+impl<'a> AsFd for io::StderrLock<'a> {
+    #[inline]
+    fn as_fd(&self) -> BorrowedFd<'_> {
+        unsafe { BorrowedFd::borrow_raw(2) }
```
Comment (joshtriplett) [inline library/std/src/sys/wasi/stdio.rs:126]:
```suggestion
        // SAFETY: user code should not close stderr out from under the standard library
        unsafe { BorrowedFd::borrow_raw(2) }
```

File: library/std/src/sys/wasi/stdio.rs (hunk 1 above)
Comment (joshtriplett) [reply library/std/src/sys/wasi/stdio.rs:37]: I agree, but note that the implementations of `AsFd` for UNIX don't have such comments either. Nonetheless, added as a suggestion.

**Comment by joshtriplett** [comment]
@sunfishcode Your commit is entitled "Import types, and add stability attributes.", but I don't see any imports added.

File: library/std/src/sys/wasi/stdio.rs (hunk 1 above)
Comment (sunfishcode) [reply library/std/src/sys/wasi/stdio.rs:38]: It's a little tricky. WASI itself doesn't make such a guarantees (Unix doesn't either, for that matter). It's efffectively Rust's std that makes this guarantee/requirement, and for its part, its own API doesn't provide a way to close stdin/stdout/stderr without using `unsafe`. @joshtriplett's added comments look ok for now, though eventually we should add public documentation about this.

**Comment by sunfishcode** [comment]
> @sunfishcode Your commit is entitled "Import types, and add stability attributes.", but I don't see any imports added.

Oh, my first patch added `use crate::io::StdinLock;`, but then I noticed that the Unix counterpart uses `io::StdinLock` etc., so I updated it to match, and didn't update the commit message. I'll amend the commit message.

**Pull request merged**